Readme.md
toml.py
------

Python parser for [mojombo/toml](https://github.com/mojombo/toml).

Improved version of [marksteve/toml-ply](https://github.com/marksteve/toml-ply)

[![Build Status](https://travis-ci.org/hit9/toml.py.png?branch=master)](https://travis-ci.org/hit9/toml.py)

Support
-------

Version [V0.1](https://github.com/mojombo/toml/blob/master/versions/toml-v0.1.0.md) of toml.

**Python>=2.5 or >=3.3**

Install
-------

    pip install toml.py

Parser
------

```python
>>> import toml
>>> toml.loads("""
... [blog]
... [blog.author]
... name = "Tom"
... age = 14
... score = 9.99
... is_child = true
... """)
{'blog': {'author': {'age': 14, 'score': 9.99, 'name': 'Tom', 'is_child': True}}}
```

Unicode string are also supported:

```python
>>> toml.loads(u"""
... name = "汤姆"
... """)
{u'name': u'\u6c64\u59c6'}
```

Empty input:(only spaces, tabs or newlines)

```python
>>> toml.loads(" ")
{}
```

input only `#` comments:

```python
>>> toml.loads("# only comments!")
{}
```

from command line:

```
$ echo "n = 1.3" | python -m toml
{'n': 1.3}

```

Generator
---------

```python
>>> import toml
>>> print toml.dumps({'blog': {'author': {'age': 14, 'score': 9.99, 'name': 'Tom', 'is_child': True}}})
[blog.author]
age = 14
score = 9.99
name = "Tom"
is_child = true
```

Exception
----------

The only exception may occur during parsing: `TomlSyntaxError`.

Specific Notes
--------------

**Each piece of notes bellow comes from [mojombo/toml](https://github.com/mojombo/toml)**, just implemented in toml.py

1. Negative integer and float is ok: `-1` `-0.9`, but positive integer or float in this format is not allowed: `+9` `+8.8`

2. Boolens are always lowercase.

3. Arrays also ignore newlines between the brackets:

   ```python
   >>> import toml
   >>> toml.loads("""
   ... arr = [
   ...     2,
   ...     3,
   ...     4,
   ... ]
   ... """)
   {'arr': [2, 3, 4]}
   ```

   As you see, terminating commas are ok before the closing bracket.

4. Arrays can be nested

   ```python
   >>> toml.loads("""
   ... arr = [[1,2,[3,4],5]]
   ... """)
   {'arr': [[1, 2, [3, 4], 5]]}
   ```

5. You don't need to specify all the superkeys if you don't want to. TOML knows how to do it for you.

   ```python
   >>> toml.loads("""
   ... [x.y.z]
   ... a = "somestr"
   ... """)
   {'x': {'y': {'z': {'a': 'somestr'}}}}
   ```

6. For float, there must be at least one number on each side of the decimal point.

   ```
   .5  # bad
   ```

7. Datetimes are ISO 8601 dates, only full zulu form is allowed.

   ```
    >>> toml.loads("""date = 1979-05-27T07:32:00Z""")
   {'date': datetime.datetime(1979, 5, 27, 7, 32)}
   ```

8. In strings, all escapes from toml 0.3.1 are supported.

   ```
   \b         - backspace       (U+0008)     [x]
   \t         - tab             (U+0009)     [x]
   \n         - linefeed        (U+000A)     [x]
   \f         - form feed       (U+000C)     [x]
   \r         - carriage return (U+000D)     [x]
   \"         - quote           (U+0022)     [x]
   \\         - backslash       (U+005C)     [x]
   \uXXXX     - unicode         (U+XXXX)     [x]
   \UXXXXXXXX - unicode         (U+XXXXXXXX) [x]
   ```
   ```python
   >>> toml.loads(r"""  # attention to the r here, use raw or unicode or \\\\
   ... path = "C:\\files\\MyPython\\"
   ... """)
   {'path': 'C:\\files\\MyPython\\'}
   ```

9. Multiline strings are supported, as well as trimming whitespace.

   ```toml
   # This will return "The quick brown fox jumps over the lazy dog."""
   key1 = """
   The quick brown \
   fox jumps over \
   the lazy dog."""
   ```

   Literal strings and multiline literal strings are supported.

   ```toml
   # What you see is what you get
   winpath  = 'C:\Users\nodejs\templates'
   winpath2 = '\\ServerX\admin$\system32\'
   quoted   = 'Tom "Dubs" Preston-Werner'
   regex    = '<\i\c*\s*>'
   
   regex2 = '''I [dw]on't need \d{2} apples'''
   lines  = '''
   The first newline is
   trimmed in raw strings.
      All other whitespace
      is preserved.
      '''
   ```

10. The spaces and tabs should be in english mode.(notes for chinese .etc users.)

11. keys can have char `?` and `#`

   ```
   >>> toml.loads(' what? = "Yeah!" ')
   {'what?': 'Yeah!'}
   ```

   ```
   >>> toml.loads(' the# = false  ')
   {'the#': False}
   ```

12. if your `keygroup` ends with no keys:

    ```
    >>> toml.loads("[NoKeysInThisGroup]")                              
    {'NoKeysInThisGroup': {}}                                          
    ```

Tests
-----

So far,  only tested for parser.

    $ nosetests -w tests

TODO
----

2. Unimplemented: Data types may not be mixed. (Later..)

3. which way is better? update already exist key or raise error instead.


changes.md
toml.py changes log
===================

Version 0.1.7
--------------

- Add support for keygroup arrays via #10


tests/tests.py
# coding=utf8
# run test for toml.py
#

from __future__ import print_function

import os
import sys
import yaml
from nose.tools import raises
from functools import wraps

sys.path.insert(0, "..")
import toml
from toml import TomlSyntaxError as e

import sys

if sys.hexversion < 0x03000000:
    PY_VERSION = 2
else:
    PY_VERSION = 3

def readf(name):
    return open(os.path.join("tomls", name + ".toml")).read()


def t(func):
    # get item's name   test_string => string
    name = func.__name__[5:]

    @wraps(func)
    def _t():
        # read file by filename
        dct = toml.loads(readf(name))
        return func(dct)
    return _t


def u(func):
    # unicode
    name = func.__name__[5:]

    @wraps(func)
    def _u():
        if PY_VERSION == 2:
            dct = toml.loads(readf(name).decode("utf8"))
        else:
            dct = toml.loads(readf(name))
        return func(dct)
    return _u


@t
def test_comment(dct):
    assert dct == {}


@t
def test_key_value(dct):
    assert dct == {"name": "Tom", "age": 10}


@t
def test_ignore_tab_space(dct):
    assert dct == {"name": "mark", "email": "mark@github.com", "score": 9.8}


@t
def test_string(dct):
    assert dct == dict(
        a="Yes,  I am a string",
        b="\b",
        c="\t",
        d="\n",
        e="\f",
        f="\r",
        g="\"",
        h="\\"
    )

@t
def test_multiline_string(dct):
    assert dct == dict(
        a="Yes,\nI am a string",
        b="\b",
        c="\t",
        d="\n",
        e="\f",
        f="\r",
        g="\"",
        h="\\",
        i="TestTest"
    )

@t
def test_literal_string(dct):
    assert dct == dict(
        key1=r"C:\Users\nodejs\templates",
        key2=r"\\ServerX\admin$\system32",
        key3="Tom \"Dubs\" Preston-Werner",
        key4=r"<\i\c*\s*>"
    )

@t
def test_multiline_literal_string(dct):
    assert dct == dict(
        regex2=r"I [dw]on't need \d{2} apples",
        lines="The first newline is\ntrimmed in raw strings.\n   All other whitespace\n   is preserved.\n"
    )

@t
def test_float(dct):
    assert dct == dict(
        a=1.0,
        b=0.3,
        c=0.5,
        d=8.8,
        e=-9.4,
        f=11.6
    )


@t
def test_integer(dct):
    assert dct == dict(
        a=1,
        b=999,
        c=-888,
        d=5
    )


@raises(e)
@t
def test_bad_float_1():
    pass


@raises(e)
@t
def test_bad_float_2():
    pass


@raises(e)
@t
def test_bad_integer():
    pass


@u
def test_unicode(dct):
    assert dct == {u"name": u"你好!", u"title": u"汤姆"}

@u
def test_unicode_escape(dct):
    assert dct == {u"a": u"é", u"b": u"é"}

@t
def test_chinese(dct):
    assert dct == dict(
        name="小明",
        email="xiaoming@126.com"
    )


@t
def test_boolen(dct):
    assert dct == {"do_u_love_me": False, "see": True}


@t
def test_empty(dct):
    assert dct == {}


@t
def test_datetime(dct):
    from datetime import datetime
    assert dct == dict(
        datetime=datetime(1979, 5, 27, 7, 32)
    )


@t
def test_keygroup(dct):
    assert dct == {
        "a": {
            "name": "Toml",
            "email": "Toml@github.com",
            "age": 78,
            "b": {
                "m": 13.9
            }
        },
        "b": {
            "c": {
                "d": {
                    "e": -0.999
                }
            }
        }
    }


@t
def test_array(dct):
    assert dct == {
        "a": ["你", "是", "谁"],
        "b": [1, 2, 3, ]
    }


@t
def test_empty_keygroup(dct):
    assert dct == {
        "a": {
            "b": {
                "c": {}
            }
        },
        "c": {
            "d": {}
        }
    }

@t
def test_keygroup_array(dct):
    assert dct == {
        "a": [  { "b": "abc" },
                { "b": "def" } ]
    }

########################################
# official test suite
########################################
@t
def test_hard_example(dct):
    print("Official test suite: hard_example.toml")
    assert dct == yaml.load(open("yaml/hard_example.yaml").read())


@t
def test_example(dct):
    print("Official test suite: example.toml")
    assert dct == yaml.load(open("yaml/example.yaml").read())


toml.py
# coding=utf8
#
# Python parser for Toml(https://github.com/mojombo/toml)
# An improved version of \
# toml-ply(https://github.com/marksteve/toml-ply)
# License: MIT
# Author: hit9
#

__version__ = '0.1'  # Current supported Toml's version

import sys
from ply import lex
from ply import yacc
from six import string_types
from datetime import datetime

DATETIME_ISO8601_FORMAT = "%Y-%m-%dT%H:%M:%SZ"

if sys.version_info >= (3, 0):
    PY_VERSION = 3
else:
    PY_VERSION = 2


def _unicode_escape_(s):
    if(PY_VERSION != 2):
        s = bytes(s, 'utf-8')
    return s.decode('unicode-escape')


class TomlSyntaxError(SyntaxError):
    pass


class TomlLexer(object):

    tokens = (
        "BOOLEN",
        "KEY",
        "KEYGROUP",
        "KEYGROUPARRAY",
        "EQUALS",
        "DATETIME",
        "STRING",
        "FLOAT",
        "INTEGER",
    )

    states = (
        ('string', 'exclusive'),
        ('mulstr', 'exclusive'),
        ('literal', 'exclusive'),
        ('mulliteral', 'exclusive'),
        ('escape', 'exclusive'),
    )

    literals = "[],"

    t_ignore = "\x20\x09"  # ignore space(x20) and tab(x09)
    t_string_mulstr_literal_mulliteral_escape_ignore = ""
    t_ignore_COMMENT = r'\#.*'  # comments
    t_EQUALS = r'='

    def t_BOOLEN(self, t):
        r'true | false'
        t.value = (t.value == "true")
        return t

    def t_KEY(self, t):
        r'[a-zA-Z_][a-zA-Z0-9_#\?]*'
        return t

    def t_KEYGROUPARRAY(self, t):
        r'\[\[([a-zA-Z_][a-zA-Z0-9_#\?]*\.?)+\]\]'
        t.value = tuple(t.value[2:-2].split('.'))  # cast to group
        return t


    def t_KEYGROUP(self, t):
        r'\[([a-zA-Z_][a-zA-Z0-9_#\?]*\.?)+\]'
        t.value = tuple(t.value[1:-1].split('.'))  # cast to group
        return t

    def t_DATETIME(self, t):
        # ISO 8601 dates: 1979-05-27T07:32:00Z
        r'\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z'
        t.value = datetime.strptime(t.value, DATETIME_ISO8601_FORMAT)
        return t

    # String handling: multiline strings
    def t_triquote(self, t):
        r'\"\"\"\n?'
        t.lexer.lineno += t.value.count("\n")
        t.lexer.helperstr = ""
        t.lexer.push_state('mulstr')

    def t_mulstr_triquote(self, t):
        r'\"\"\"'
        t.value = t.lexer.helperstr
        t.type = "STRING"
        t.lexer.pop_state()
        return t

    def t_mulstr_mulliteral_newline(self, t):
        r'\n'
        t.lexer.lineno += 1
        t.lexer.helperstr += "\n"

    # This needs to be before escape characters to ensure it is handled first
    def t_mulstr_trimmedwhitespace(self, t):
        r'\\[ \n\t]+'
        t.lexer.lineno += t.value.count("\n")
        pass

    # String handling: strings
    def t_quote(self, t):
        r'\"'
        t.lexer.helperstr = ""
        t.lexer.push_state('string')

    def t_string_quote(self, t):
        r'\"'
        t.value = t.lexer.helperstr
        t.type = "STRING"
        t.lexer.pop_state()
        return t

    def t_string_literal_newline(self, t):
        r'\n'
        t.lexer.begin('INITIAL')    # Reset state so next parse works
        raise TomlSyntaxError(
            "Illegal newline in string at Line %d" % t.lineno
        )

    # String handling: multiline literal strings
    def t_trisinglequote(self, t):
        r'\'\'\'\n?'
        t.lexer.lineno += t.value.count("\n")
        t.lexer.helperstr = ""
        t.lexer.push_state('mulliteral')

    def t_mulliteral_trisinglequote(self, t):
        r'\'\'\''
        t.value = t.lexer.helperstr
        t.type = "STRING"
        t.lexer.pop_state()
        return t

    # String handling: literal strings
    def t_singlequote(self, t):
        r'\''
        t.lexer.helperstr = ""
        t.lexer.push_state('literal')

    def t_literal_singlequote(self, t):
        r'\''
        t.value = t.lexer.helperstr
        t.type = "STRING"
        t.lexer.pop_state()
        return t

    # Escape characters
    def t_string_mulstr_backslash(self, t):
        r'\\'
        t.lexer.push_state('escape')

    def t_escape_backspace(self, t):
        r'b'
        t.lexer.helperstr += "\b"
        t.lexer.pop_state()

    def t_escape_tab(self, t):
        r't'
        t.lexer.helperstr += "\t"
        t.lexer.pop_state()

    def t_escape_linefeed(self, t):
        r'n'
        t.lexer.helperstr += "\n"
        t.lexer.pop_state()

    def t_escape_form_feed(self, t):
        r'f'
        t.lexer.helperstr += "\f"
        t.lexer.pop_state()

    def t_escape_carriage_return(self, t):
        r'r'
        t.lexer.helperstr += "\r"
        t.lexer.pop_state()

    def t_escape_quote(self, t):
        r'\"'
        t.lexer.helperstr += "\""
        t.lexer.pop_state()

    def t_escape_backslash(self, t):
        r'\\'
        t.lexer.helperstr += "\\"
        t.lexer.pop_state()

    def t_escape_unicode(self, t):
        r'(u[a-fA-F0-9]{4,4}|U[a-fA-F0-9]{8,8})'
        u = _unicode_escape_("\\" + t.value)
        t.lexer.helperstr += u
        t.lexer.pop_state()

    # Catch all rule needs to be the last to avoid conflict with previous
    # more specific rules
    def t_string_mulstr_literal_mulliteral_char(self, t):
        r'.'
        t.lexer.helperstr += t.value

    def t_FLOAT(self, t):
        r'([-]?(\d+)(\.\d+)(e(\+|-)?(\d+))? | (\d+)e(\+|-)?(\d+))([lL]|[fF])?'
        t.value = float(t.value)
        return t

    def t_INTEGER(self, t):
        r'[-]?\d+([uU]|[lL]|[uU][lL]|[lL][uU])?'
        t.value = int(t.value)
        return t

    def t_newline(self, t):
        r'\n+'
        t.lexer.lineno += len(t.value)

    def t_ANY_error(self, t):
        t.lexer.begin('INITIAL')    # Reset state so next parse works
        raise TomlSyntaxError(
            "Illegal character: '%s' at Line %d" % (t.value[0], t.lineno)
        )

    def __init__(self):
        self.lexer = lex.lex(module=self)


class TomlParser(object):

    tokens = TomlLexer.tokens

    def p_error(self, p):
        if p:
            raise TomlSyntaxError(
                "Character '%s' at line %d, \
                token: %r" % (p.value[0], p.lineno, p)
            )
        else:
            raise TomlSyntaxError("SyntaxError at EOF")

    def p_start(self, p):
        # parser start here
        "start : translation_unit"
        p[0] = self.dct

    def p_translation_unit(self, p):
        # unit to find out all assignments
        """
        translation_unit : assignment
                         | translation_unit assignment
                         |
        """
        pass

    def p_assignment_key(self, p):
        # look up all keys
        """
        assignment : KEY EQUALS value
        """
        d = self.dct
        # slide to the current keygroup's depth
        for key in self.keygroup:
            d = d[key]

        if isinstance(d,list):
            d = d[-1]

        # if duplicate, recover the old one
        # But I really dont know how to raise an error here
        # raise statement seems not working here
        d[p[1]] = p[3]

    def p_assignment_keygroup_array(self, p):
        # look up all keygroups
        """
        assignment : KEYGROUPARRAY
                   | assignment KEYGROUPARRAY
        """

        self.keygroup = p[len(p) - 1]
        listname = self.keygroup[-1]

        d = self.dct

        for key in self.keygroup[:-1]:
            # init the keygroup's value to empty dict
            d = d.setdefault(key, {})

        # if it already exists, then insert a new array
        mylist = d.get( listname )
        if not mylist:
            mylist = d.setdefault(listname, [] )

        mylist.append({})


    def p_assignment_keygroup(self, p):
        # look up all keygroups
        """
        assignment : KEYGROUP
                   | assignment KEYGROUP
        """
        self.keygroup = p[len(p) - 1]
        d = self.dct

        for key in self.keygroup:
            # init the keygroup's value to empty dict
            d = d.setdefault(key, {})

    def p_value(self, p):
        # values can be array, int, datetime, float, string integer, boolen
        """
        value : array
              | BOOLEN
              | DATETIME
              | STRING
              | FLOAT
              | INTEGER
        """

        p[0] = p[1]

    def p_array(self, p):
        """array : '[' sequence ']'"""
        p[0] = p[2]

    def p_sequence(self, p):
        # sequence is: a, b, c, d, ..
        """
        sequence : sequence ',' value
                 | sequence ','
                 | value
                 |
        """

        if len(p) == 1:
            p[0] = []
        elif len(p) == 2:
            p[0] = [p[1]]
        elif len(p) == 3:
            p[0] = p[1]
        elif len(p) == 4:
            p[0] = p[1] + [p[3]]

    def __init__(self):
        self.parser = yacc.yacc(module=self, debug=0, write_tables=0)

    def parse(self, toml_str):
        # reset dct and keygroup
        self.dct = dict()  # dict object for parser to return
        self.keygroup = tuple()  # to record current group
        return self.parser.parse(toml_str)


class TomlGenerator(object):  # generate toml string from valid python dict

    g_newline = "\n"

    def g_string(self, v):
        # annoying escaping chars :)
        o = ""
        esc_chars = {
            "\t": "t",
            "\n": "n",
            "\"": '"',
            "\r": "r",
            "\\": "\\",
            "\f": "f",
            "\b": "b",
        }

        for c in v:
            if c in esc_chars:
                o += "\\" + esc_chars[c]
            else:
                o += c

        return '"' + o + '"'

    def g_bool(self, v):
        return "true" if v else "false"

    def g_integer(self, v):
        return str(v)

    def g_float(self, v):
        return str(v)

    def g_datetime(self, v):
        return v.strftime(DATETIME_ISO8601_FORMAT)

    def g_array(self, v):
        lst = [self.gen_value(i) for i in v]
        return "[" + ", ".join(lst) + "]"

    def gen_value(self, v):
        # generate toml format of python data
        if isinstance(v, string_types):
            return self.g_string(v)
        elif isinstance(v, bool):
            return self.g_bool(v)
        elif isinstance(v, int):
            return self.g_integer(v)
        elif isinstance(v, float):
            return self.g_float(v)
        elif isinstance(v, datetime):
            return self.g_datetime(v)
        elif isinstance(v, list):
            return self.g_array(v)
        else:
            raise TomlSyntaxError("Invalid data type: %r" % (type(v), ))

    def gen_section(self, dct, keygroup):
        section, body = [], []

        for key, value in dct.items():
            if isinstance(value, dict):
                section.append(self.gen_section(value, keygroup + [key]))
            else:
                body.append(key + " = " + self.gen_value(value))

        if body and keygroup:
            body.insert(0, "[" + ".".join(keygroup) + "]")

        return self.g_newline.join(body + section)


lexer = TomlLexer()  # build lexer
parser = TomlParser()  # build parser
generator = TomlGenerator()  # init a Generator instance


def loads(toml_str):
    return parser.parse(toml_str)


def dumps(dct):
    return generator.gen_section(dct, [])

if __name__ == '__main__':
    if PY_VERSION == 2:
        input_string = raw_input()
    elif PY_VERSION == 3:
        input_string = input()
    exit(loads(input_string))


toooml.py
# coding=utf8
#
# Python parser for toml (Tom's Obvious, Minimal Language)
# Home: https://github.com/mojombo/toml
# Note: This is an improved version of \
# toml-ply(https://github.com/marksteve/toml-ply)
# License: MIT
#

__version__ = '0.1'

from ply import lex
from ply import yacc
from ply.lex import TOKEN
from ply.yacc import YaccProduction
from datetime import datetime
from re import UNICODE

# escape string
# \b     - backspace       (U+0008)
# \t     - tab             (U+0009)
# \n     - linefeed        (U+000A)
# \f     - form feed       (U+000C)
# \r     - carriage return (U+000D)
# \"     - quote           (U+0022)
# \/     - slash           (U+002F)
# \\     - backslash       (U+005C)
# \uXXXX - unicode         (U+XXXX)
# But i dont think toml should escape this char: /
# see mojombo/toml/issue#173. I dont want to escape forward slashes
ES = r"(\\([btnfr\"\\u]|[0-7]{1,3}|x[a-fA-F0-9]+))"

# string's literal
STR = r'\"([^"\\\n]|' + ES + ')*\"'


class TomlSyntaxError(SyntaxError):
    pass


ISO8601 = "%Y-%m-%dT%H:%M:%SZ"

tokens = (
    "BOOLEN",
    "KEY",
    "KEYGROUP",
    "EQUALS",
    "DATETIME",
    "STRING",
    "FLOAT",
    "INTEGER",
)

literals = ["[", "]", ","]

# ignore space(x20) and tab(x09)
t_ignore = "\x20\x09"

# comments
# key = "value" # Yeah, you can do this.
t_ignore_COMMENT = r'\#.*'


def t_newline(t):
    r'\n+'
    t.lexer.lineno += len(t.value)


def t_error(t):
    raise TomlSyntaxError(
        "Illegal character at %r" % (t, )
    )


t_EQUALS = r"="


def t_BOOLEN(t):
    r'true|false'
    t.value = (t.value == "true")
    return t


def t_KEY(t):
    r'[a-zA-Z_][a-zA-Z0-9_]*'
    return t


# keygroups can be nested. so use tuple to store this.
# e.g. [a.b.c] => ('a', 'b', 'c')
def t_KEYGROUP(t):
    r'\[([a-zA-Z_][a-zA-Z0-9_]*\.?)+\]'
    t.value = tuple(t.value[1:-1].split('.'))
    return t


# ISO 8601 dates: 1979-05-27T07:32:00Z
def t_DATETIME(t):
    r'\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z'
    t.value = datetime.strptime(t.value, ISO8601)
    return t


@TOKEN(STR)
def t_STRING(t):
    # remove fisrt double quote and last double quote as value
    t.value = t.value[1:-1]
    return t


# Feeling negative? Do what's natural.
# toml means, don't use this: +1.79, use 1.79 instead
def t_FLOAT(t):
    r'([-](\d+)(\.\d+)(e(\+|-)?(\d+))? | (\d+)e(\+|-)?(\d+))([lL]|[fF])?'
    t.value = float(t.value)
    return t


# dont use +4,  use 4 instead.
# negative integer is ok: -4
def t_INTEGER(t):
    r'[-]?\d+([uU]|[lL]|[uU][lL]|[lL][uU])?'
    t.value = int(t.value)
    return t


# build lexer
lex.lex(reflags=UNICODE)

# return dict
dct = None
# temp global var to record keygroup
keygroup = None


def p_error(p):
    raise TomlSyntaxError(
        "SyntaxError at %r" % (p, )
    )

# start rule, store dct
def p_start(p):
    "start : translation_unit"
    p[0] = dct


# unit to lookup all assignments
def p_translation_unit(p):
    """
    translation_unit : assignment
                     | translation_unit assignment
    """
    pass


# lookup all keys
def p_assignment_keys(p):
    """assignment : KEY EQUALS value"""
    # TODO: if key already in dct, raise error
    # reference global dct with d
    d = dct
    # if keys are in some keygroup
    for k in keygroup:
        # @marksteve 's way is right!
        d.setdefault(k, {})
        d = d[k]
    d[p[1]] = p[3]


# looup all keygroups
def p_assignment_keygroup(p):
    """assignment : KEYGROUP
                  | assignment KEYGROUP"""
    global keygroup
    keygroup = p[len(p)-1]

# values can be array, int, datetime, float, string integer, boolen
def p_value(p):
    """value : array
             | BOOLEN
             | DATETIME
             | STRING
             | FLOAT
             | INTEGER"""

    p[0] = p[1]


# Arrays are square brackets with other primitives inside.
# Whitespace is ignored. Elements are separated by commas.
# Data types may not be mixed.
def p_array(p):
    """array : '[' sequence ']'"""
    p[0] = p[2]


# terminating commas are ok before the closing bracket.
def p_sequence(p):
    """sequence : value
                | sequence ','
                | sequence ',' value"""

    if len(p) == 1:
        p[0] = []
    elif len(p) == 2 or len(p) == 3:
        p[0] = [p[1]]
    elif len(p) == 4:
        p[0] = p[1] + [p[3]]


parser = yacc.yacc(debug=0, write_tables=0)


def loads(s):
    global dct
    global keygroup
    # reset return dict
    dct = dict()
    keygroup = tuple()
    if s:
        return parser.parse(s)
    # return {} if s is empty
    return {}

if __name__ == '__main__':
    exit(loads(raw_input()))
